sphinx/api/schema.rst

```diff
@@ -0,0 +1,295 @@
+the following subsections, in addition to the frequency (:py:attr:`freq`) for time-dependent inputs. All
+other attributes of the metadata classes for user reference, and therefore immutable. After setting
+each of the inputs, users can access the dictionary elements :py:attr:`col_map`, :py:attr:`dtypes`, and :py:attr:`units` to
+work with the various mappings. Below, is a demonstration of this mapping in practice, showing the
+SCADA data mapping used in "examples/data/plant_meta.yml", where the keys are the OpenOA column names,
+and the values are the La Haute Borne data naming conventions. This mapping can be repeated for each
+of the other metadata types.
+
+.. literalinclude:: ../../examples/data/plant_meta.yml
+    :language: yaml
+    :lines: 37-45
+    :linenos:
+    :lineno-start: 37
+
+
+
+Data Schema User Guide
+**********************
+
+The following subsections will demonstrate the required data mapping schemas to enable
+:py:class:`openoa.plant.PlantData` to validate and and convert user-specified data to a validated
+OpenOA schema for use throughout the codebase. The data columns and their associated units and
+datatypes will be shown in a table, followed by a demonstration of how this is used in the La Haute
+Borne example data used for all of the example analysis workflows. It should be noted that the
+column "Field Name" is the internal naming convention, and should be the dictionary or JSON/YAML
+key with the actual column naming as its associated value (as is seen in the YAML snippets for each
+section).
+
+It should be noted though, that validating a :py:class:`PlantData` object with
+:py:attr:`analysis_type` = "all" will check for all of the Field Names listed below for all provided
+data. However, if the :py:class:`PlantData` object is only being validated for a
+specific analysis type, or types, then only the data specified in
+:py:const:`openoa.plant.ANALYSIS_REQUIREMENTS` (shown below) will be checked, and in the case of
+:py:attr:`analysis_type` = None, then no errors will be raised during validation.
+
+.. literalinclude:: ../../openoa/schema/metadata.py
+    :language: python
+    :lines: 24-73
+    :linenos:
+    :lineno-start: 24
+
+SCADA
+=====
+
+:py:attr:`PlantData.scada` is configured by the :py:class:`openoa.schema.SCADAMetaData` class, which is set in the configuration
+data with the "scada" key. Users can set each of the following "Field Name" keys with their own
+data's column names in their SCADA data, plus the "freq" field. Users just have to ensure that the
+columns are already using the specified units, and that each column is already using the listed
+data type or can be converted to that type.
+
+==================== ================= ================================== =============================
+ Field Name          Descriptive Name   Data Type (SCADAMetaData.dtypes)   Units (SCADAMetaData.units)
```

Comment:
This table is hard-coded in the RST documentation, whereas the YAML below is automatically generated from the python code. Is there a plan to ensure these remain synchronized?

Reply:
For now this will stay a manual process unless we have a good way to generate and insert an RST table into the documentation page. The difficulty here is that we're not including the more descriptive types of information in the metadata classes, so it can't be generated from anywhere. Given that this is part of the more descriptive aspect of the documentation, and not the autogenerated material from docstrings, I'm ok with not solving this problem for now, especially if we don't foresee that many changes to the schema.

As for an assurance the remain synced, it could be added to a PR checklist or something somewhere, but I don't think we really have anything like that set up here anyway.

Reply:
For your first question, the tests should catch that the schema files and the analysis requirements schema don't match, which would indicate that we need to regenerate the schema files, which is just running `python openoa/schema/schema.py`. One thing I like about it not being auto-generated is that now we have to stop and see what's changed, and forces us to actual address the above as well because our schema has changed. That said, this is definitely an imperfect solution, but I don't have a great resolution to solve it all in one fell swoop at the moment, especially if this might change in the future before the full v3 release.

I think also that the JSON/YAML files should stay because it gives people a quick, in-code reference for the data without having to read the code or the documentation. Essentially I'm thinking of it as a redundant form of documentation.

Reply:
For #229, the proposed workflow wouldn't be possible because the plant data object has to be created first. The key feature that is likely missing here is further defining the flairs of each base analysis class, so instead of "MonteCarloAEP" only, we could also have something like "MonteCarloAEP-temp" or "MonteCarloAEP-winddirection" that each add in a variation of requirements from the base analysis type. Currently these are only handled by the analysis class.